File: src/backend/src/Dealmatcher.Backend.Domain/EntityAggregates/UserAggregate/Ban.cs
namespace Dealmatcher.Backend.Domain.EntityAggregates.UserAggregate;

public class Ban : DealmatcherEntityBase
{
    public User User { get; private set; } = null!;
    public string Reason { get; private set; } = null!;
    public User IssuedBy { get; private set; } = null!;
    public DateTime IssuedAt { get; private set; }
    public DateTime? ExpiresAt { get; private set; }
    public bool IsActive { get; private set; }

    /* EF Core*/
    private Ban() { }

    internal Ban(User user, string reason, User issuedBy, DateTime? expiresAt)
    {
        User = user;
        Reason = reason;
        IssuedBy = issuedBy;
        IssuedAt = DateTime.UtcNow;
        ExpiresAt = expiresAt;
        IsActive = true;
    }

    internal void Revoke()
    {
        IsActive = false;
    }
}

File: src/frontend/lib/models/ban.dart
class Ban {
  const Ban({
    required this.id,
    required this.userId,
    required this.reason,
    required this.issuedBy,
    required this.issuedAt,
    required this.expiresAt,
    required this.isActive,
  });

  final int id;
  final int userId;
  final String reason;
  final int issuedBy;
  final DateTime issuedAt;
  final DateTime? expiresAt;
  final bool isActive;

  factory Ban.fromJson(Map<String, dynamic> json) {
    return Ban(
      id: json["id"],
      userId: json["userId"],
      reason: json["reason"],
      issuedBy: json["issuedBy"],
      issuedAt: DateTime.parse(json["issuedAt"]),
      expiresAt: json["expiresAt"] == null
          ? null
          : DateTime.parse(json["expiresAt"]),
      isActive: json["isActive"],
    );
  }
}
